### ISSUE TITLE
brew upgrade -g errors on linux

### ISSUE THREAD

**Issue by adamhl8** [Homebrew/brew#16135]
### `brew doctor` output

```shell
Please note that these warnings are just used to help the Homebrew maintainers
with debugging if you file an issue. If everything you use Homebrew for is
working fine: please don't worry or file an issue; just ignore this. Thanks!

Warning: Putting non-prefixed coreutils in your path can cause GMP builds to fail.
```

### Verification

- [X] My "`brew doctor` output" above says `Your system is ready to brew.` and am still able to reproduce my issue.
- [X] I ran `brew update` twice and am still able to reproduce my issue.
- [X] This issue's title and/or description do not reference a single formula e.g. `brew install wget`. If they do, open an issue at https://github.com/Homebrew/homebrew-core/issues/new/choose instead.

### `brew config` output

```shell
HOMEBREW_VERSION: 4.1.17
ORIGIN: https://github.com/Homebrew/brew
HEAD: 35746e0a6ba6c3c5cfe56d99f79d9ec9f52ee15f
Last commit: 2 days ago
Core tap JSON: 24 Oct 18:37 UTC
HOMEBREW_PREFIX: /home/linuxbrew/.linuxbrew
HOMEBREW_CASK_OPTS: []
HOMEBREW_EDITOR: micro
HOMEBREW_MAKE_JOBS: 10
Homebrew Ruby: 2.6.10 => /home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/vendor/portable-ruby/2.6.10_1/bin/ruby
CPU: 10-core 64-bit unknown_0x6_0xa7
Clang: N/A
Git: 2.42.0 => /home/linuxbrew/.linuxbrew/bin/git
Curl: 8.4.0 => /bin/curl
Kernel: Linux 6.5.0-2-amd64 x86_64 GNU/Linux
OS: Debian GNU/Linux trixie/sid (trixie)
Host glibc: 2.37
/usr/bin/gcc: 13.2.0
/usr/bin/ruby: N/A
glibc: N/A
gcc@11: N/A
gcc: 13.2.0
xorg: N/A
```

### What were you trying to do (and why)?

Trying to run `brew upgrade -g` on a linux host. I realize that casks aren't supported on linux, but I'm running this as part of a larger script that I share between different machines.

### What happened (include all command output)?

`Error: Invalid usage: Options --formula and --greedy are mutually exclusive.`

### What did you expect to happen?

I would hope that brew would effectively ignore the `-g` option if on a host that does not support casks or if there are no casks installed.

### Step-by-step reproduction instructions (by running `brew` commands)

```shell
brew upgrade -g
```

**Comment by apainintheneck** [comment]
Passing `--formula` by default for non-dev commands on Linux was something we added a while ago in https://github.com/Homebrew/brew/pull/14206 to help reduce confusion for Linux users. Basically, it wasn't as clear why the package wouldn't install correctly (https://github.com/orgs/Homebrew/discussions/3687). We eventually allowed it for dev commands to allow some commands for tap management to be used on Linux runners in CI (https://github.com/Homebrew/brew/pull/14344).

The solution was admittedly very hacky so it's not super easy to modify.

My first thought is that you could probably do something like this:

```rb
system("brew upgrade --formula")
system("brew upgrade --cask -g") if RUBY_PLATFORM.include?("darwin")
```

Basically, run the command first to upgrade formula and then upgrade casks if you're on macOS. I assume that should be relatively straightforward with whatever language you're using for your script.

If that doesn't work, you could share a snippet or a brief overview of what you're trying to accomplish with your script.

**Comment by MikeMcQuaid** [comment]
> My first thought is that you could probably do something like this:

That's the best idea here.

---

The error message could be better here but ultimately what you're asking for is a behaviour that doesn't exist on Linux so you should not pass it.

Another option would be to set the `HOMEBREW_UPGRADE_GREEDY` environment variable only on macOS.

**Comment by adamhl8** [comment]
> My first thought is that you could probably do something like this:

Yeah that's basically what I'm doing in my script. Which is fine, just wasn't sure if the error I was getting was intended behavior. I appreciate the responses.
